test/integration/plowing_effect.cc

```diff
@@ -0,0 +1,197 @@
+ private: void CallbackRigidTerrain(const ConstContactsPtr &_msg);
+ private: void CallbackDeformableTerrain(const ConstContactsPtr &_msg);
+
+ private: physics::CollisionPtr wheelCollisionPtr_ = nullptr;
+
+  /// constant params from SDF
+ private: double maxRadians_;
+ private: double deadbandVelocity_;
+ private: double saturationVelocity_;
+ private: ignition::math::Vector3<double> fdir1_;
+};
+
+////////////////////////////////////////////////////////////////////////
+void PlowingEffect::CallbackRigidTerrain(const ConstContactsPtr &_msg)
+{
+  std::string wheelCollisionStr = "original_tricycle::wheel_front::collision";
+  std::string groundCollisionStr = "plowing_effect_ground_plane::link::collision";
+
+  // no shift in contact point on rigid surface
+  for(auto idx = 0; idx < _msg->contact_size(); ++idx)
+  {
+    const gazebo::msgs::Contact& contact = _msg->contact(idx);
+    if(contact.collision1() == wheelCollisionStr &&
+       contact.collision2() == groundCollisionStr)
+    {
+      const ignition::math::Vector3<double>& contactPointNormal =
+                            ConvertIgn( contact.normal(0));
+      const ignition::math::Vector3<double>& unitNormal =
+                            ignition::math::Vector3<double>::UnitZ;
+      double angle = acos(contactPointNormal.Dot(unitNormal)/
+                     contactPointNormal.Length() * unitNormal.Length());
+      ASSERT_EQ(angle, 0);
+    }
+  }
+}
+
+void PlowingEffect::CallbackDeformableTerrain(const ConstContactsPtr &_msg)
+{
+  std::string wheelCollisionStr = "my_tricycle::wheel_front::collision";
+  std::string groundCollisionStr = "plowing_effect_ground_plane::link::collision";
+
+  for(auto idx = 0; idx < _msg->contact_size(); ++idx)
+  {
+    const gazebo::msgs::Contact& contact = _msg->contact(idx);
+    if(contact.collision1() == wheelCollisionStr &&
+        contact.collision2() == groundCollisionStr)
+    {
+      const auto& contactPointNormal = ConvertIgn( contact.normal(0));
+      const auto& unitNormal = ignition::math::Vector3<double>::UnitZ;
+
+      // compute longitudinal unit vector as normal cross fdir1_
+      const auto& wheelLinearVelocity = wheelCollisionPtr_->WorldLinearVel();
+      const auto& unitLongitudinal = contactPointNormal.Cross(fdir1_);
+      double wheelSpeedLongitudinal = wheelLinearVelocity.AbsDot(unitLongitudinal);
+      
+      // compute the angle contact point makes with unit normal
+      double angle = acos(contactPointNormal.Dot(unitNormal)/contactPointNormal.Length() * unitNormal.Length());
+
+      // no plowing effect
+      if(wheelSpeedLongitudinal < deadbandVelocity_)
+      {
+        EXPECT_EQ(angle, 0);
+      }
+
+      // plowing effect
+      else if(wheelSpeedLongitudinal > deadbandVelocity_ && wheelSpeedLongitudinal < saturationVelocity_)
+      {
+          // angle less than maxRadians_
+          ASSERT_LT(angle, maxRadians_);
+        // TODO: should we also check for exact angle values here
```

Comment:
@adityapande-1995 for this should I compute the exact angle using the same logic used in ODEPhysics?

Reply:
We could calculate the angle and use EXPECT_NEAR macro to test, by including some tolerance.